fxa-oauth-server/lib/routes/authorization.js

```diff
@@ -218,104 +128,102 @@ module.exports = {
+      logger.critical('joi.response_type', { response_type: req.payload.response_type });
+      throw AppError.invalidResponseType();
+    }
+  }
+};
 
-        var uri = req.payload.redirect_uri || client.redirectUri;
 
-        if (uri !== client.redirectUri) {
-          logger.debug('redirect.mismatch', {
-            param: uri,
-            registered: client.redirectUri
-          });
+async function generateAuthorizationCode(client, payload, grant) {
+  // Clients must use PKCE if and only if they are a pubic client.
+  if (client.publicClient) {
+    if (! payload.code_challenge_method || ! payload.code_challenge) {
+      logger.info('client.missingPkceParameters');
+      throw AppError.missingPkceParameters();
+    }
+  } else {
+    if (payload.code_challenge_method || payload.code_challenge) {
+      logger.info('client.notPublicClient');
+      throw AppError.notPublicClient({ id: payload.client_id });
+    }
+  }
 
-          if (config.get('localRedirects') && isLocalHost(uri)) {
-            logger.debug('redirect.local', { uri: uri });
-          } else {
-            throw AppError.incorrectRedirect(uri);
-          }
+  const state = payload.state;
 
-        }
+  let code = await db.generateCode(Object.assign(grant, {
+    codeChallengeMethod: payload.code_challenge_method,
+    codeChallenge: payload.code_challenge,
+  }));
+  code = hex(code);
 
-        if (wantsGrant && ! client.canGrant) {
-          logger.warn('implicitGrant.notAllowed', {
-            id: req.payload.client_id
-          });
-          throw AppError.invalidResponseType();
-        }
+  const redirect = URI(payload.redirect_uri).addQuery({ code, state });
 
-        req.payload.redirect_uri = uri;
+  return {
+    code,
+    state,
+    redirect: String(redirect)
+  };
+}
 
-        checkPKCEParams(req, client);
+// N.B. We do not correctly implement the "implicit grant" flow from
+// RFC6749 which defines `response_type=token`. Instead we have a
+// privileged set of clients that use `response_type=token` for something
+// approximating the "resource owner password grant" flow, using an identity
+// assertion to just directly grant tokens for their own use. Known current
+// users of this functinality include:
+//
+//  * Firefox Desktop, for getting "profile"-scoped tokens to access profile data
+//  * Firefox for Android, for getting "profile"-scoped tokens to access profile data
+//  * Firefox for iOS, for getting "profile"-scoped tokens to access profile data
+//
+// New clients should not do this, and should instead of `grant_type=fxa-assertion`
+// on the /token endpoint.
+//
+// This route is kept for backwards-compatibility only.
+async function generateImplicitGrant(client, payload, grant) {
+  if (! client.canGrant) {
+    logger.warn('grantType.notAllowed', {
+      id: hex(client.id),
+      grant_type: 'fxa-assertion'
+    });
+    throw AppError.invalidResponseType();
+  }
+  return await generateTokens(Object.assign(grant, {
```

Comment:
If generateTokens returns a promise, you shouldn't need the `await` here do you?

Reply:
Correct, I think I mostly do this out of habit